src/tla/Apalache.tla

```diff
@@ -1,103 +1,109 @@
-(*                                                                           *)
-(* For a further discussion on that matter, see:                             *)
-(* https://github.com/informalsystems/apalache/blob/ik/idiomatic-tla/docs/idiomatic/assignments.md *)
-(*****************************************************************************)
+(**
+ * An assignment of an expression e to a state variable x. Typically, one
+ * uses the non-primed version of x in the initializing predicate Init and
+ * the primed version of x (that is, x') in the transition predicate Next.
+ * Although TLA+ does not have a concept of a variable assignment, we find
+ * this concept extremely useful for symbolic model checking. In pure TLA+,
+ * one would simply write x = e, or x \in {e}.
+ *
+ * Apalache automatically converts some expressions of the form
+ * x = e or x \in {e} into assignments. However, if you like to annotate
+ * assignments by hand, you can use this operator.
+ *
+ * For a further discussion on that matter, see:
+ * https://github.com/informalsystems/apalache/blob/ik/idiomatic-tla/docs/idiomatic/assignments.md
+ *)
 x := e == x = e
 
-(*****************************************************************************)
-(* A generator of a data structure. Given a positive integer `bound`, and    *)
-(* assuming that the type of the operator application is known, we           *)
-(* recursively generate a TLA+ data structure as a tree, whose width is      *)
-(* bound by the number `bound`.                                              *)
-(*                                                                           *)
-(* The body of this operator is redefined by Apalache.                       *)
-(*****************************************************************************)
+(**
+ * A generator of a data structure. Given a positive integer `bound`, and
+ * assuming that the type of the operator application is known, we
+ * recursively generate a TLA+ data structure as a tree, whose width is
+ * bound by the number `bound`.
+ *
+ * The body of this operator is redefined by Apalache.
+ *)
 Gen(size) == {}
 
-(*****************************************************************************)
-(* As TLA+ is untyped, one can use function- and sequence-specific operators *)
-(* interchangeably. However, to maintain correctness w.r.t. our type-system, *)
-(* an explicit cast is needed when using functions as sequences.             *)
-(*****************************************************************************)
+(**
+ * Convert a set of pairs S to a function F. Note that if S contains at least
+ * two pairs <<x, y>> and <<u, v>> such that x = u and y /= v,
+ * then F is not uniquely defined. We use CHOOSE to resolve this ambiguity.
+ * Apalache implements a more efficient encoding of this operator
+ * than the default one.
+ *
+ * @type: Set(<<a, b>>) => (a -> b);
+ *)
+SetAsFun(S) ==
+    LET Dom == { x: <<x, y>> \in S }
+        Rng == { y: <<x, y>> \in S }
```

Comment:
Not a criticism of your code, but wow, this syntax sugar is horrible. The fact that you're allowed to pattern-match on pairs, when `S` could be any set is outright strange.

Reply:
Yes, but it works in TLA+

Reply:
I know, that's the weird part.